plugins/Compressor/Compressor.cpp

```diff
@@ -0,0 +1,630 @@
+	const bool autoMakeup = m_compressorControls.m_autoMakeupModel.value();
+	const int stereoLink = m_compressorControls.m_stereoLinkModel.value();
+	const bool audition = m_compressorControls.m_auditionModel.value();
+	const bool feedback = m_compressorControls.m_feedbackModel.value();
+	const bool lookahead = m_compressorControls.m_lookaheadModel.value();
+
+	for(fpp_t f = 0; f < frames; ++f)
+	{
+		sample_t drySignal[2] = {buf[f][0] * m_inGainVal, buf[f][1] * m_inGainVal};
+		sample_t s[2] = {drySignal[0], drySignal[1]};
+
+		// Calculate tilt filters, to bias the sidechain to the low or high frequencies
+		if (m_tiltVal)
+		{
+			calcTiltFilter(s[0], s[0], 0);
+			calcTiltFilter(s[1], s[1], 1);
+		}
+
+		s[0] *= inBalance > 0 ? 1 - inBalance : 1;
+		s[1] *= inBalance < 0 ? 1 + inBalance : 1;
+
+		float gainResult[2] = {0, 0};
+
+		for (int i = 0; i < 2; i++)
+		{
+			float inputValue = feedback ? m_prevOut[i] : s[i];
+
+			// Calculate the crest factor of the audio by diving the peak by the RMS
+			m_crestPeakVal[i] = qMax(inputValue * inputValue, m_crestTimeConst * m_crestPeakVal[i] + (1 - m_crestTimeConst) * (inputValue * inputValue));
+			m_crestRmsVal[i] = m_crestTimeConst * m_crestRmsVal[i] + ((1 - m_crestTimeConst) * (inputValue * inputValue));
+			m_crestFactorVal[i] = m_crestPeakVal[i] / m_crestRmsVal[i];
+
+			// Grab the peak or RMS value
+			inputValue = qMax(COMP_NOISE_FLOOR, peakmode ? abs(inputValue) : m_rms[i]->update(inputValue));
+
+			// The following code uses math magic to semi-efficiently
+			// find the largest value in the lookahead buffer.
+			// This can probably be improved.
+			if (lookahead)
+			{
+				// Pre-lookahead delay, so the total delay always matches 20 ms
+				++m_preLookaheadBufLoc[i];
+				if (m_preLookaheadBufLoc[i] >= m_preLookaheadLength)
+				{
+					m_preLookaheadBufLoc[i] = 0;
+				}
+				const float tempInputValue = inputValue;
+				inputValue = m_preLookaheadBuf[i][m_preLookaheadBufLoc[i]];
+				m_preLookaheadBuf[i][m_preLookaheadBufLoc[i]] = tempInputValue;
+
+
+				// Increment ring buffer location
+				++m_lookaheadBufLoc[i];
+				if (m_lookaheadBufLoc[i] >= m_lookaheadLength)
+				{
+					m_lookaheadBufLoc[i] = 0;
+				}
+
+				m_lookaheadBuf[i][m_lookaheadBufLoc[i]] = inputValue;
+
+				// If the new input value is larger than the stored maximum,
+				// store that as the maximum
+				if (inputValue >= m_maxLookaheadVal[i])
+				{
+					m_maxLookaheadVal[i] = inputValue;
+					m_maxLookaheadTimer[i] = m_lookaheadLength;
+				}
+
+				// Decrement timer.  When the timer reaches 0, that means the
+				// stored maximum value has left the buffer and a new
+				// maximum value must be found.
+				if (--m_maxLookaheadTimer[i] <= 0)
+				{
+					m_maxLookaheadTimer[i] = std::distance(std::begin(m_lookaheadBuf[i]),
+						std::max_element(std::begin(m_lookaheadBuf[i]), std::begin(m_lookaheadBuf[i]) + m_lookaheadLength));
```

Comment:
Note, it's possible to optimize finding the maximum value in a ring buffer using a double-ended queue. I can provide the detail if you want.
BTW, why do you want to use the maximum value in the lookahead buffer?

Reply:
It seems like this doesn't consume CPU much, so I think the existing comment is enough which says about the possibility of further optimization.

Reply:
> Note, it's possible to optimize finding the maximum value in a ring buffer using a double-ended queue. I can provide the detail if you want.

Would that be faster than O(n)? 😮

Reply:
It's amortized O(1) for fixed lookahead value, and additional O(log(n)) if it's automated.